data-jdbc/src/main/java/io/micronaut/data/jdbc/operations/DefaultJdbcRepositoryOperations.java

```diff
@@ -864,155 +869,211 @@ private void cascadeUpdate(Class<?> repositoryType, AnnotationMetadata annotatio
-                        }
-                        if (association.isForeignKey()) {
-                            association.getInverseSide().ifPresent(inverse -> {
-                                final BeanProperty property = inverse.getProperty();
-                                property.set(associated, entity);
-                            });
-                        }
-
-                        // get the insert operation
-                        StoredInsert<Object>  associatedInsert = resolveEntityInsert(
+                        cascadePersistForChild(connection, annotationMetadata, repositoryType, entity, association, entry.getValue(), persisted);
+                        break;
+                    case ONE_TO_MANY:
+                    case MANY_TO_MANY:
+                    default:
+                        cascadePersistForChildren(connection,
                                 annotationMetadata,
                                 repositoryType,
-                                associationType,
-                                associatedEntity
-                        );
+                                entity,
+                                insert.getPersistentEntity(),
+                                identityProperty,
+                                insert.getDialect(),
+                                association,
+                                persisted);
+                }
+            }
+        }
+    }
 
-                        if (associatedId != null) {
-                            final Object id = associatedIdProperty.get(associated);
-                            if (id != null) {
-                                continue;
-                            }
-                        }
+    private void cascadePersistForChild(Connection connection,
+                                        AnnotationMetadata annotationMetadata,
+                                        Class<?> repositoryType,
+                                        Object parent,
+                                        RuntimeAssociation<Object> childAssociation,
+                                        Object child,
+                                        Set<Object> persisted) {
+        final RuntimePersistentEntity<?> associatedEntity = childAssociation.getAssociatedEntity();
+        final Class<?> associationType = associatedEntity.getIntrospection().getBeanType();
+        final RuntimePersistentProperty<?> associatedId = associatedEntity.getIdentity();
+        final BeanProperty associatedIdProperty = associatedId.getProperty();
+        // in the case of an one-to-one we ensure the inverse side set
+        if (child == null || persisted.contains(child)) {
+            return;
+        }
+        if (childAssociation.isForeignKey()) {
+            childAssociation.getInverseSide().ifPresent(inverse -> {
+                final BeanProperty property = inverse.getProperty();
+                property.set(child, parent);
```

Comment:
What if the property is read-only? Maybe `withValue` and copy constructor approach as fallback?

Reply:
I need to add more records tests to check all the scenarios.